terrain-manager.js

```diff
@@ -0,0 +1,180 @@
+		this.positionAttribute.setUsage( THREE.DynamicDrawUsage );
+
+		this.normalAttribute = new THREE.Float32BufferAttribute();
+		this.normalAttribute.array = this.bufferFactory.normals;
+		this.normalAttribute.itemSize = 3;
+		this.normalAttribute.count = this.bufferFactory.normals.length / 3;
+		this.normalAttribute.setUsage( THREE.DynamicDrawUsage );
+
+		this.geometry.setIndex(this.indexAttribute);
+		this.geometry.setAttribute('position', this.positionAttribute);
+		this.geometry.setAttribute('normal', this.normalAttribute);
+
+		this.geometry.clearGroups();
+
+		for (let i = 0; i < this.chunkCount * this.chunkCount; i++) {
+			this.geometry.addGroup(
+				this.bufferFactory.indexRanges[2 * i], this.bufferFactory.indexRanges[2 * i + 1], 0
+			);
+		}
+
+		this.mesh = new THREE.Mesh(
+			this.geometry, [new THREE.MeshLambertMaterial({ color: 0xff0000, wireframe: true })]
+		);
+	}
+
+	_calculateTargetChunks() {
+
+		let centerChunkGridX = Math.floor(this.center.x / this.chunkSize);
+		let centerChunkGridZ = Math.floor(this.center.z / this.chunkSize);
+
+		let targetChunks = [];
+
+		for (let i = centerChunkGridX - this.chunkRange; i < centerChunkGridX + this.chunkRange + 1; i++) {
+			for (let j = centerChunkGridZ - this.chunkRange; j < centerChunkGridZ + this.chunkRange + 1; j++) {
+				targetChunks.push(i + ':' + j);
+			}
+		}
+
+		return targetChunks;
+	}
+
+	updateCenter(pos) {
+
+		this.center = pos;
+
+		this.targetChunkIds = this._calculateTargetChunks();
+	}
+
+	updateChunk() {
+
+		const buf = this.bufferFactory;
+
+		let chunkIdToAdd = this.targetChunkIds.filter(
+			id => !this.currentChunks.map(v => v.chunkId).includes(id)
+		).at(0);
+
+		console.log(">>> chunk id to add: ", chunkIdToAdd);
+
+		let chunksToRemove = this.currentChunks.filter(chunk => !this.targetChunkIds.includes(chunk.chunkId));
+
+		console.log(">>> chunk id to remove: ", chunksToRemove);
+
+		chunksToRemove.forEach(chunk => {
+			this.geometryUtils.deallocateChunk(
+				chunk.slots[0], chunk.slots[1], this.chunkCount ** 2,
+				buf.chunkVertexRangeBuffer,
+				buf.vertexFreeRangeBuffer,
+				buf.chunkIndexRangeBuffer,
+				buf.indexFreeRangeBuffer
+			);
+		});
+
+		this.currentChunks = this.currentChunks.filter(chunk => this.targetChunkIds.includes(chunk.chunkId));
+
+		if (!!chunkIdToAdd) {
+			let gridId = chunkIdToAdd.split(':');
+
+			let slots = this.geometryUtils.allocateChunk(
+				buf.positionBuffer, buf.normalBuffer, buf.indexBuffer,
+				buf.chunkVertexRangeBuffer,
+				buf.vertexFreeRangeBuffer,
+				buf.chunkIndexRangeBuffer,
+				buf.indexFreeRangeBuffer,
+				gridId[0] * this.chunkSize, 0, gridId[1] * this.chunkSize,
+				this.chunkSize, this.segment, this.chunkCount ** 2
+			);
+
+			this.currentChunks.push({ slots: slots, chunkId: chunkIdToAdd });
+
+			this._updateChunkGeometry(slots);
+		}
+
+	}
+
+	_updateChunkGeometry(slots) {
+
+		const buf = this.bufferFactory;
+
+		this.indexAttribute.updateRange = {
```

Comment:
This isn't going to work when we update more than one chunk per frame.

We will need to manually call the buffer attribute function when we want to update:

https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLAttributes.js#L143